dot/sync/chain_sync.go

```diff
@@ -53,693 +48,630 @@ func (s chainSyncState) String() string {
-}
+	logger.Debugf("💤 waiting for %d blocks", totalBlocks)
+	syncingChain := make([]*types.BlockData, totalBlocks)
+	// the total numbers of blocks is missing in the syncing chain
+	waitingBlocks := totalBlocks
 
-func (cs *chainSync) tryDispatchWorker(w *worker) {
-	// if we already have the maximum number of workers, don't dispatch another
-	if len(cs.workerState.workers) >= maxWorkers {
-		logger.Trace("reached max workers, ignoring potential work")
-		return
-	}
+	for waitingBlocks > 0 {
+		// in a case where we don't handle workers results we should check the pool
+		idleDuration := time.Minute
+		idleTimer := time.NewTimer(idleDuration)
 
-	// check current worker set for workers already working on these blocks
-	// if there are none, dispatch new worker
-	if cs.handler.hasCurrentWorker(w, cs.workerState.workers) {
-		return
-	}
+		select {
+		case <-cs.stopCh:
+			return
 
-	cs.workerState.add(w)
-	go cs.dispatchWorker(w)
-}
+		case <-idleTimer.C:
+			logger.Warnf("idle ticker triggered! checking pool")
+			cs.workerPool.useConnectedPeers()
+			continue
 
-// dispatchWorker begins making requests to the network and attempts to receive responses up until the target
-// if it fails due to any reason, it sets the worker `err` and returns
-// this function always places the worker into the `resultCh` for result handling upon return
-func (cs *chainSync) dispatchWorker(w *worker) {
-	if w.targetNumber == nil || w.startNumber == nil {
-		return
-	}
+		// TODO: implement a case to stop
+		case taskResult := <-workersResults:
+			if !idleTimer.Stop() {
+				<-idleTimer.C
+			}
 
-	logger.Debugf("dispatching sync worker id %d, "+
-		"start number %d, target number %d, "+
-		"start hash %s, target hash %s, "+
-		"request data %d, direction %s",
-		w.id,
-		*w.startNumber, *w.targetNumber,
-		w.startHash, w.targetHash,
-		w.requestData, w.direction)
+			logger.Debugf("task result: peer(%s), with error: %v, with response: %v",
+				taskResult.who, taskResult.err != nil, taskResult.response != nil)
+
+			if taskResult.err != nil {
+				logger.Errorf("task result: peer(%s) error: %s",
+					taskResult.who, taskResult.err)
+
+				if !errors.Is(taskResult.err, network.ErrReceivedEmptyMessage) {
+					switch {
+					case strings.Contains(taskResult.err.Error(), "protocols not supported"):
+						cs.network.ReportPeer(peerset.ReputationChange{
+							Value:  peerset.BadProtocolValue,
+							Reason: peerset.BadProtocolReason,
+						}, taskResult.who)
+						cs.workerPool.ignorePeerAsWorker(taskResult.who)
+					default:
+						cs.workerPool.punishPeer(taskResult.who)
+					}
+				}
 
-	start := time.Now()
-	defer func() {
-		end := time.Now()
-		w.duration = end.Sub(start)
-		outcome := "success"
-		if w.err != nil {
-			outcome = "failure"
-		}
-		logger.Debugf(
-			"sync worker completed in %s with %s for worker id %d",
-			w.duration, outcome, w.id)
-		cs.resultQueue <- w
-	}()
+				cs.workerPool.submitRequest(taskResult.request, workersResults)
```

Comment:
is this not the case where error is network.ErrReceivedEmptyMessage?

Reply:
Yeah, if we receive `network.ErrReceivedEmptyMessage` we don't punish the peer, IMO doesn't make sense to punish it for don't be able to send us a slice of the chain, so we just submit the same request for another peer